### PR TITLE
cryptenroll: use root device by default

### PR DESCRIPTION
**Pull request by lnussel**

### CODE REVIEW

File: src/cryptenroll/cryptenroll.c [hunk 1]
```diff
@@ -534,17 +536,31 @@ static int parse_argv(int argc, char *argv[]) {
...
+                        return r;
+        } else {
+                dev_t devno;
+
+                r = blockdev_get_root(LOG_ERR, &devno);
+                if (r < 0)
+                        return r;
+                if (r == 0) {
+                        log_error("Root file system not backed by a (single) whole block device.");
+                        return 80; /* some recognizable error code */
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: do you actually care about the error?

Note that you have to switch from DEFINE_MAIN_FUNCTION() to DEFINE_MAIN_FUNCTION_WITH_POSITIVE_FAILURE() in the last line of the file if this shall be propagated.

but frankly, i'd just do `return log_error_errno(SYNTHETIC_ERRNO(ENXIO), "…");` here, i.e. return a regular negative error which will then turned into EXIT_FAILURE as last step. Unless you actually really care about a recognizable error.

If you really do want the recognizable error then please try to use a BSD error, i.e. the stuff from `/usr/include/sysexits.h`, maybe EX_PROTOCOL or EX_UNAVAILABLE or so? or maybe even EX_DATAERR...

but again, i'd just not bother

**Comment by poettering** [comment]
makes sense, let's try this.

needs a man page update too. needs to document that the device is now optional, and what the effect of not specifying it is.

also, if you really want to keep the recognizable exit status, you need to document that one too.

File: src/cryptenroll/cryptenroll.c (hunk 1 above)
Comment (lnussel) [reply src/cryptenroll/cryptenroll.c]: ah, right. Too much c&p from bootctl :see_no_evil:

**Comment by AdrianVovk** [comment]
First, please see https://github.com/systemd/systemd/pull/31096/commits/b417fa11475c006bbdb734d2efabdace89e6405a for prior art :)

There's two issues I see w/ this MR as-is:

1. We shouldn't just make the device node optional outright. We should make the automatic discovery behind a switch. This is just to ensure that someone messing around w/ the cryptenroll command on the command line doesn't accidentally nuke their keys. This was a concern @poettering brought up to me when we last discussed implementing this

2. We should be finding the device that's backing /var, not the root directory. See reasoning here: https://github.com/systemd/systemd/pull/31096#discussion_r1467678983

**Comment by lnussel** [comment]
Well hidden :-) I don't mind closing this in favor of yours. Alternatively we go ahead with this and you correct the code to use /var in yours.

Requiring an extra 'auto' argument in whatever form would kind of defeat the purpose though IMO. I really like when `systemd-cryptenroll` just does something meaningful when called without arguments.

**Comment by poettering** [comment]
Oh, yikes, I knew I had seen the idea before. Sorry for not realizing where I saw it...

**Comment by AdrianVovk** [comment]
The varlink stuff won't land for a long while. So probably better to fix up this PR and land the change separately 

> Requiring an extra 'auto' argument in whatever form would kind of defeat the purpose though IMO.

This was my opinion originally too but since there's no confirmation it becomes easy to nuke your system

I think the point is more not having to figure out what block device to operate on, which can be tricky and vary between installations of a distro

**Comment by poettering** [comment]
we could probably find a middle ground regarding the auto thing, i.e. refuse to do the auto device logic of --wipe= is used without enrolling a new key. 

because --wipe= is the only strictly destructive thing, and hence the only thing we really need to be careful with (i mean, the worst thing enrolling can otherwise do is allow people in with a low quality pw, but that's not the most horrible thing in the world i'd say, you can easily recover from that, even if there's a time window where things are not as protected as they should be). 

And I'd claim that if you enroll something then it means you can still get access to the disk right after. i.e. we have no unlock credentials that you can enroll but that would then self destroy so that you couldn't back into the volume with them right after.

Hence, if we only do the auto stuff if there's actually at least one enrollment done, then we should be fine.

**Comment by poettering** [comment]
that said, I'd also be fine to say you have to add `-a` to the cmdline to enable auto mode or so.

**Comment by lnussel** [comment]
updated to only use the device automatically if no wipe operation is requested.

**Comment by lnussel** [comment]
the ci fail here seems to be unrelated

**Pull request merged**